torch/_inductor/codegen/cpp.py

```diff
@@ -3782,48 +3824,173 @@ def can_fuse_vertical(self, node1, node2):
+            """
+            assert isinstance(node, OuterLoopFusedSchedulerNode)
+            cpp_kernel_proxy_list.clear()
+            nodes_list.clear()
+
+            def get_call_ranges(node: BaseSchedulerNode):
+                assert isinstance(node, (SchedulerNode, FusedSchedulerNode))
+                nodes: List[SchedulerNode] = node.get_nodes()  # type: ignore[assignment]
+                _, (group, reduction_group) = max(
+                    nodes, key=lambda x: int(x.is_reduction())
+                ).group
+                call_ranges = tuple(group) + tuple(reduction_group)
+                return call_ranges
+
+            LocalBuffer = namedtuple("LocalBuffer", ["local_buf", "global_buf"])
+            local_buffers: List[LocalBuffer] = []
+            if all(
+                len(get_call_ranges(_node)) == node.outer_loop_fusion_depth + 1
+                for _node in node.get_outer_nodes()
             ):
-                # Merge the cpp_kernel_proxy_list into cpp_kernel_proxy
+                # Ref to the typical case of local buffer
+                # in https://github.com/pytorch/pytorch/blob/
+                # 1115a25c36340554442f28f9570abd42f0aface2/aten/src/ATen/native/cpu/SoftMaxKernel.cpp#L159
+                # where the buffer is with size of last dim and contiguous.
+                # Only support this typical case at first.
+                for scheduler_node in node.get_nodes():
+                    # all users inside same OuterLoopFusedSchedulerNode
+                    if not scheduler_node.is_reduction() and all(
+                        user.node in node.get_nodes() for user in scheduler_node.users
+                    ):
+                        global_buffer = scheduler_node.node
+                        assert isinstance(global_buffer, ir.ComputedBuffer)
+                        global_buffer_layout = global_buffer.get_layout()
+                        size_offset = node.outer_loop_fusion_depth - len(
+                            get_call_ranges(scheduler_node)
+                        )
+
+                        def is_all_write_read_contiguous(scheduler_node):
+                            def is_contiguous_index(index_expr, sizes):
+                                index = sorted(
+                                    index_expr.free_symbols, key=lambda s: s.name
+                                )
+                                return all(
+                                    stride_at(index_expr, index[idx]) == sizes[idx + 1]
+                                    if ((idx + 1) < len(sizes))
+                                    else 1
+                                    for idx in range(len(sizes))
+                                )
+
+                            write_index_expr = scheduler_node._body.writes_name2expr[
+                                scheduler_node.get_name()
+                            ]
+                            sizes = scheduler_node._sizes[0]
```

Comment:
I'm not sure your expected stride values are correct and in general `stride_at` is not a sound way to interpret index expressions because we can have things like `ModularIndex`. Would something like this work instead.
```suggestion
                            contiguous_index_expr = 0
                            stride = 1
                            for var, range in reversed(scheduler_node._body.var_ranges.items()):
                                contiguous_index_expr += stride * var
                                stride *= range

                            write_index_expr = scheduler_node._body.writes_name2expr[
                                scheduler_node.get_name()
                            ]
                            
                            def is_contiguous_index(x):
                                return x == contiguous_index_expr
```

Reply:
Thanks a lot for the suggestion! Changed as it.